test/functional/p2p_orphan_handling.py

```diff
@@ -0,0 +1,416 @@
+        # Relay orphan child_B
+        self.relay_transaction(peer_orphans, child_B["tx"])
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        # Only missing_parent_B should be requested. Not inflight_parent_AB or missing_parent_AB
+        # because they are already being requested from peer_txrequest and peer_orphans respectively.
+        peer_orphans.wait_for_parent_requests([int(missing_parent_B["txid"], 16)])
+        peer_orphans.assert_never_requested(int(inflight_parent_AB["txid"], 16))
+
+    @cleanup
+    def test_orphan_of_orphan(self):
+        node = self.nodes[0]
+        peer = node.add_p2p_connection(PeerTxRelayer())
+
+        self.log.info("Test handling of an orphan with a parent who is another orphan")
+        missing_grandparent = self.wallet_nonsegwit.create_self_transfer()
+        missing_parent_orphan = self.wallet_nonsegwit.create_self_transfer(utxo_to_spend=missing_grandparent["new_utxo"])
+        missing_parent = self.wallet_nonsegwit.create_self_transfer()
+        orphan = self.wallet_nonsegwit.create_self_transfer_multi(utxos_to_spend=[missing_parent["new_utxo"], missing_parent_orphan["new_utxo"]])
+
+        # The node should put missing_parent_orphan into the orphanage and request missing_grandparent
+        self.relay_transaction(peer, missing_parent_orphan["tx"])
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        peer.wait_for_parent_requests([int(missing_grandparent["txid"], 16)])
+
+        # The node should put the orphan into the orphanage and request missing_parent, skipping
+        # missing_parent_orphan because it already has it in the orphanage.
+        self.relay_transaction(peer, orphan["tx"])
+        self.nodes[0].bumpmocktime(NONPREF_PEER_TX_DELAY + TXID_RELAY_DELAY)
+        peer.wait_for_parent_requests([int(missing_parent["txid"], 16)])
+
+    @cleanup
+    def test_orphan_inherit_rejection(self):
+        node = self.nodes[0]
+        peer1 = node.add_p2p_connection(PeerTxRelayer())
+        peer2 = node.add_p2p_connection(PeerTxRelayer())
+        peer3 = node.add_p2p_connection(PeerTxRelayer())
+
+        self.log.info("Test that an orphan with rejected parents, along with any descendants, cannot be retried with an alternate witness")
+        parent_low_fee_nonsegwit = self.wallet_nonsegwit.create_self_transfer(fee_rate=0)
+        assert_equal(parent_low_fee_nonsegwit["txid"], parent_low_fee_nonsegwit["tx"].getwtxid())
+        child = self.wallet.create_self_transfer(utxo_to_spend=parent_low_fee_nonsegwit["new_utxo"])
+        grandchild = self.wallet.create_self_transfer(utxo_to_spend=child["new_utxo"])
+        assert child["txid"] != child["tx"].getwtxid()
+        assert grandchild["txid"] != grandchild["tx"].getwtxid()
+
+        # Relay the parent. It should be rejected because it pays 0 fees.
+        self.relay_transaction(peer1, parent_low_fee_nonsegwit["tx"])
```

Comment:
Could check that `parent_low_fee_nonsegwit` is not in the mempool as done in the other test cases.